teps/0121-refine-retries-for-taskruns-and-customruns.md

```diff
@@ -129,14 +129,36 @@ Task-level Timeout (`TaskRunSpec.Timeout` and `RunSpec.Timeout`) is set for each
 
 ### Retries in TaskRuns and CustomRuns
 
-Add a new `Retries` field to [TaskRunSpec](https://pkg.go.dev/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1#TaskRunSpec). Model the `CustomRunSpec` based on the `RunSpec` to use the existing `retries` field.
+Add a new `Retries` field in [TaskRunSpec](https://pkg.go.dev/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1#TaskRunSpec). Model the `CustomRunSpec` based on the `RunSpec` to use the existing `retries` field.
 
 The `PipelineTask.Retries` value, which is specified at `Pipeline` authoring time, is passed to the `TaskRunSpec.Retries` and `CustomRunSpec.Retries` during the execution of a `PipelineRun`.
 
 The `TaskRun` and `CustomRun` controllers handle their own `Retries`. 
 
 The `PipelineRun` controller does not check for `len(retriesStatus)` to determine whether a `TaskRun` or `CustomRun` is done executing. Instead, uses `ConditionSucceeded` as the only way to decide if the `TaskRun` or `CustomRun` has completed execution.
 
+#### Execution Status of a `failed` `pipelineTask` with `retries`
+
+The following table shows how an overall status of a `taskRun` or a `customRun` for a `pipelineTask` with `retries` set to 3:
+
+| `status` | `reason`    |`completionTime` is set | description                                                                                                |
+|----------|-------------|------------------------|------------------------------------------------------------------------------------------------------------|
+| Unknown  | Running     | No                     | The `taskRun` has been validated and started to perform its work.                                          |
+| Unknown  | ToBeRetried | No                     | The `taskRun` (zero attempt of a `pipelineTask`) finished executing but failed. It has 3 retries pending.  |
```

Comment:
I understand the purpose of this intermediate `reason` but I'm not sure how it will be implemented.
When the `TaskRun` reconciler detects a failure in the underlying `Pod`, it will:

- store the data of the current run in `retriesStatus`
- if there are no retries left, mark the `TaskRun` as failed
- else, clean up the current `TaskRun` status
- set the `reason` to `ToBeRetried`

At this point, the controller could start the next `TaskRun`, but if it does so, the `ToBeRetried` value will never make it into `etcd`. To make the `ToBeRetried` reason visible to watchers we will have to finish the reconcile cycle and start the retry in the next reconcile cycle.

Is this the planned implementation? It seems to me that the reconcile cycle should do ask much as possible to reconcile the resource, rather than stop and wait for the next reconcile. 
I'm guessing that the changes to the `TaskRun` done by the reconcile will trigger another reconcile cycle, but I'm not 100% sure.